Payment Buttons: Add justification, orientation and block-spacing controls

dsas opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to projects/plugins/jetpack/to-test.md in a new commit as part of your PR. -->
<!-- a12s: If you have an expected version that you're aiming for the PR to add, please use the Milestone field to communicate it. If you leave it blank, that indicates there isn't a preference. -->

Fixes #25927

#### Changes proposed in this Pull Request:
<!--- Explain what functional changes your PR includes -->
Add grid layout options to Payment Buttons, making the design just as configurable as Core Buttons. Specifically this brings the following controls:
 * Justification
 * Orientation
 * Block spacing

#### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?


#### Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->

#### Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->

#### Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

* Create a [JN](https://href.li/?https://jurassic.ninja/create/?jetpack-beta&branch=master) / wpcom site with a paid plan
* Insert some Payment Blocks, go through the stripe connection etc
* Publish the post and open it in a new window
* Use JP beta to switch your site to this branch / apply this patch to your sandbox
* Open the published post in a new window and compare the two - they should look identical
* Open the published post in the editor, it should load as normal - no errors
* Twiddle with the new options: the preview should update accordingly and after saving so should the published post.

Before | After
-------|------
![image](https://user-images.githubusercontent.com/93301/190361773-e600ca4a-610c-4792-9f79-0433c17e3c3d.png) | ![image](https://user-images.githubusercontent.com/93301/190361841-9704e355-cd33-4ca8-86a7-aa245b62feba.png)

dsas committed:
[not verified] blockGap working but not button widths

dsas committed:
[not verified] Remove some unused rules

dsas committed:
[not verified] Make button widths work

Make button widths work - they change the size of the buttons, but it's
done based upon --jetpack-payment-buttons-gap (0.5em) rather than the
actual block gap in use, so it isn't 100% correct yet.

Buttons widths are always in percents in a multipe of 25, so it can
hardcode a certain number of calculations. Payment buttons allow widths
of any number in percents, pixels and ems.

dsas committed:
[not verified] changelog

dsas commented:
Payment Button width controls work, but the button widths are not calculated in the same way are core button widths are. For a visual difference. I have a Payment Buttons block with an 11px blockGap and three buttons at 50%, 25%, 25% and have set-up a core Buttons block in the same way. It looks like this:
![image](https://user-images.githubusercontent.com/93301/189364338-4e494fdd-38fb-4d45-b64b-469913cb0fd2.png)

~~[Core button widths are sized to exclude block spacing](https://github.com/WordPress/gutenberg/blob/0ac7078173b519fbfcc2e5a623d610cf5668e189/packages/block-library/src/button/style.scss#L56-L71), that is a width of 50% is actually 50% - the blockGap. Buttons widths are always in percents in a multiple of 25, so it can hardcode a certain number of calculations.~~

~~Payment buttons allow widths of any number in percents, pixels and ems. So a slightly
different approach to upstream is in order. It's less complicated than appears at first glance, as unless you using % based block widths, you get exactly what you ask for.~~

~~recurring-payments buttons have some JS which sets styles based upon `--jetpack-payment-buttons-gap` (0.5em) rather than the actual block gap in use when using percentage based widths. I just need to adapt this to get the selected blockGap.~~ 

I'd misunderstood that.

[Core button widths are sized to exclude block spacing](https://github.com/WordPress/gutenberg/blob/0ac7078173b519fbfcc2e5a623d610cf5668e189/packages/block-library/src/button/style.scss#L56-L71), but assume block spacing is `--wp--style--block-gap` (1.5 rem) falling back to 0.5ems. It doesn't use the currently set `blockGap`. Buttons widths are always in percents in a multiple of 25, so it can hardcode a certain number of calculations, which only works when the blockGap is not set, or is smaller than the above.

Payment buttons allow widths of any number in percents, pixels and ems, so adopting hardcoded calculations for specific button widths isn't as simple as in core/buttons.

recurring-payments buttons has some JS which is used with percentage based widths to set styles based upon `--jetpack-payment-buttons-gap` (0.5em) rather than based upon the theme, or configured blockGap.

Visual differences between core/buttons and payment-buttons are solely due to the different gap used. PBs use `--jetpack-payment-buttons-gap` (0.5em) while core/buttons uses 1.5rem. I'm not planning on solving the difference here to bring them in-line or to use the configured blockGap because:
 * Changing the `--jetpack-payment-buttons-gap` to match core/buttons will change how already existing created payment buttons look
 * Accounting for the configured blockGap is horrendously complicated, and ought to be solved in core first/simultaneously

dsas committed:
[not verified] Ensure server side of block knows of layout support

This means that Gutenberg's layout.php will markup the view output with
the appropriate classes and styles.

mmtr commented on projects/plugins/jetpack/extensions/blocks/payment-buttons/edit.js:
```diff
@@ -65,18 +65,16 @@ function PaymentButtonsEdit( { clientId } ) {
 
 	return (
-		<div { ...blockProps }>
+		<>
```
This makes the Stripe nudge and the Jetpack upgrade nudge to be out of the block edit component, causing an overlap between the nudges and the block toolbar:

<img width="700" alt="Screen Shot 2022-09-15 at 13 43 09" src="https://user-images.githubusercontent.com/1233880/190395131-1bb2310d-badb-43d9-bf1c-121adbe121ac.png">

<img width="913" alt="Screen Shot 2022-09-15 at 13 44 37" src="https://user-images.githubusercontent.com/1233880/190395335-46b7d3b8-50d4-444b-8d1e-13cd5ce6f3d4.png">

dsas replied:
Hmm, the reason I made this change was because the flex container would then contain the div that wraps all of the buttons, rather than the buttons directly, setting a width on the buttons then sets the wrong width, see below. With the nudge it wraps both the nudge and the button container.

![image](https://user-images.githubusercontent.com/93301/190422509-28b11567-d8c1-4742-a4e6-164de0b436e8.png)

Seems that with the nudge, this approach isn't viable 👎 I'll have to think of something else.

mmtr replied:
Oooh, I see – that's unfortunate.

> I'll have to think of something else.

If that proves to be complex (something that I kinda expect), we have the alternative approach of moving the nudges to the block edit component after it has been rendered, which is how we add the upgrade nudges to the paid blocks in WP.com:

https://github.com/Automattic/jetpack/blob/28ca04ee102e28945c99ecaed3a8122d4709dde5/projects/plugins/jetpack/extensions/extended-blocks/paid-blocks/with-upgrade-banner.jsx#L74-L100

dsas replied:
Oh of course the wpcom and jp implementations of upgrade nudges are completely separate 😬

dsas replied:
For wpcom this is broken too albeit in a slightly different way, for basically the same reason - the outer div had a display:flex rule thanks to the is-layout-grid class.  ![image](https://user-images.githubusercontent.com/93301/190619014-24553b3d-81de-438f-8f58-a7398ff1652b.png)

The menu ultimately works by looking for the div with the clientId, so we have to put the upgrade nudges inside that div.

The layouts work [at edit time](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-editor/src/hooks/layout.js#L384-L388) by applying the `is-layout-flex` css class to the `BlockListBlock` (i.e. the component rendered on canvas) and [at view time](https://github.com/WordPress/gutenberg/blob/trunk/lib/block-supports/layout.php#L383-L390) by adding the same class name to the outermost render content.

I think the solution is to change the edit to go back to `<div { ...blockProps }>` but to delete blockProps.className first. This way the display:flex won't be applied in the wrong place. I'll give this a try.

dsas replied:
Seems to work fine 👍

mmtr replied:
> For wpcom this is broken too albeit in a slightly different way, for basically the same reason - the outer div had a display:flex rule thanks to the is-layout-grid class.

I still reproduce the same, not sure if I'm missing something but I made sure that my sandbox is using the most recent changes

<img width="674" alt="Screen Shot 2022-09-16 at 16 32 14" src="https://user-images.githubusercontent.com/1233880/190663678-89bb426c-2d26-4bfa-a72b-ec36d52ef58e.png">

dsas replied:
You're absolutely right, it's still broken on wpcom. I don't really understand why, for some reason the is-layout-flex (and other classes) are still being applied to the outer div on wpcom but not on jp.

Also both the outer and list blocks div get the same id, so if trying to make this approach work, that needs addressing too.

dsas replied:
This is basically the same problem as we had before in #25102 - the wpcom upgrade nudge hooks itself inside the edit component and we can't shift it elsewhere as we're not in control of it.

I don't think this can fundamentally work without first changing how the upgrade nudge works. 😞

dsas replied:
I've addressed this issue by creating a div for the wpcom upgrade nudge to be placed in, rather than allowing the wpcom code to create the nudge div where it thinks is best.

It works, I _think_ it's ok but please let me know if this is too much of a hack. Ideally I think the wpcom nudge would work rather differently.

mmtr reviewed (commented):
This looks pretty good! It tests very well except for a regression to the placement of the banners inside the block.

dsas committed:
[not verified] Fix upgrade nudges in editor

The upgrade nudges were appearing beneath the block toolbar menu. This
was because the outer div had a display:flex rule thanks to the is-layout-grid
class.

The layouts work at edit time by applying the is-layout-flex css class
to the BlockListBlock (i.e. the component rendered on canvas) and at
view time by adding the same class name to the outermost rendered content.

This commit changes the edit component to go back to <div { ...blockProps }>
but to delete blockProps.className first. This way the `display:flex` won't be
applied in the wrong place.

dsas committed:
[not verified] inline supports

dsas commented:
One thing to note is that for people on wpcom who have payment-buttons but who don't have the right plan to use them, the blocks will not display correctly when viewing on the front-end.

![image](https://user-images.githubusercontent.com/93301/190661132-0818b78f-2676-4ff1-8312-f93163a04935.png)

This is because [the block won't be registered with PHP](https://github.com/Automattic/jetpack//blob/01980e3a1ff8a5c841cbf687fc40d2c4f378e20c/projects/plugins/jetpack/extensions/blocks/payment-buttons/payment-buttons.php#L28-L55) and so the [__experimentalLayout support will not be available](https://github.com/WordPress/gutenberg/blob/trunk/lib/block-supports/layout.php#L289) when rendering in php.

I _think_ this behaviour is ok, but I'm definitely open to challenge on it

dsas committed:
Merge remote-tracking branch 'origin/trunk' into update/payment-buttons-layout

dsas committed:
Fix display of wpcom nudges

WPcom nudges were including the nudge as part of the button grid. This
change addresses that by adding HTML data attributes and classnames,
along with a div, so that the wpcom nudge code uses the div controlled
by this block, rather than by introducing its own, where it guesses is
best.

dsas committed:
Merge remote-tracking branch 'origin/trunk' into update/payment-buttons-layout

mmtr commented on projects/plugins/jetpack/extensions/blocks/payment-buttons/payment-buttons.php:
```diff
@@ -28,7 +28,18 @@ function register_block() {
-			array( 'render_callback' => __NAMESPACE__ . '\render_block' )
+			array(
+				'render_callback' => __NAMESPACE__ . '\render_block',
+				'supports'        => array(
```
Do we really need to declare the layout support in the backend as well?

dsas replied:
Yes, because that's how some of the css works in Gutenberg at view time. It uses the [`gutenberg_render_layout_support_flag` PHP function ](https://github.com/WordPress/gutenberg/blob/0b30a4cb34d39c9627b6a3795a18aee21019ce25/lib/block-supports/layout.php#L294) PHP function to register some styles for a `wp-container-N` class and updates the output block to include that classname. Without support declaration that function exits early.

Absence of this for people without a valid wpcom plan causes the [display to be incorrect](https://github.com/Automattic/jetpack/pull/26134#issuecomment-1249424791)

mmtr replied:
I wonder how the Core Buttons block works then, since it doesn't declare any `__experimentalLayout` support in the backend. Perhaps [the block.json file](https://github.com/WordPress/gutenberg/blob/0b30a4cb34d39c9627b6a3795a18aee21019ce25/packages/block-library/src/buttons/block.json#L34) is used by both the frontend and the backend?

Anyway, not a blocker for this.

dsas replied:
That's exactly what happens, moving the JP blocks to use block.json seemed a little too far out of scope for this PR, maybe worth doing another time though.

mmtr commented on projects/plugins/jetpack/extensions/blocks/payment-buttons/edit.js:
```diff
@@ -65,18 +65,23 @@ function PaymentButtonsEdit( { clientId } ) {
 
+	// The ID needs to be just on the outermost wrapper - the toolbar and wpcom upgrade nudge
+	// will then be positioned in relation to this.
+	delete innerBlocksProps.id;
```
The upgrade nudge does not use uses the `id` attribute to determine its position. Why do we need to delete it?

dsas replied:
I think it still works without deleting it from the inner block, but not deleting it means that there are two elements (inner and outer) with the same HTML id, which violates the HTML spec. Also it loses symmetry with data-block (which is a prefixed version of the id)

mmtr replied:
Oh right, I see it now. It's a side effect of using separately `blockProps` and `innerBlocksProps`, since `innerBlocksProps` extends from `blockProps`.

mmtr commented on projects/plugins/jetpack/extensions/blocks/payment-buttons/edit.js:
```diff
@@ -65,18 +65,23 @@ function PaymentButtonsEdit( { clientId } ) {
+	// will then be positioned in relation to this.
+	delete innerBlocksProps.id;
+	delete innerBlocksProps[ 'data-block' ];
+	// The class may include flex positioning, this must only be on the inner block wrapper
+	// otherwise the jetpack/wpcom upgrade nudge will be displayed as part of the flex
+	delete blockProps.className;
```
Do you think this could be achieved with CSS by overriding the class to unset the flex styles? Just in case the class also defines non-flex styles that would be missing otherwise.

dsas replied:
hmm, yes, I could do something like:
```css
.wp-block-jetpack-payment-buttons.is-layout-flex {
    display: block;
}

.wp-block-jetpack-payment-buttons.is-layout-flex .is-layout-flex {
    display: flex;
}
```

I'd rather not have an exhaustive list of styles to unset, I i don't think any of the current remaining is-layout-flex styles matter on display:block anyway

mmtr commented on projects/plugins/jetpack/extensions/blocks/payment-buttons/edit.js:
```diff
@@ -105,6 +110,8 @@ function PaymentButtonsEdit( { clientId } ) {
 					</div>
 				</Placeholder>
 			) }
+			{ /* Explicitly position the wpcom upgrade nudge wrapper */ }
+			<div className="jetpack-block-upgrade-banner-container"></div>
```
The wrapper is added as the first children of the block container dynamically which should result in this same markup, so you shouldn't need to explicitly add this div here. Doesn't that work?

dsas replied:
Yeah good catch, I think this is left over from an earlier, discarded attempt. I should make this change.

mmtr reviewed (commented):
Awesome work, @dsas! It works perfectly now. I just left some minor questions.

dsas committed:
Rather than removing all classes, use CSS to just change the styles used for the outer container

dsas committed:
Remove pre-positioning of the upgrade nudge - not necessary

mmtr reviewed (approved):
💥

Copons commented:
Please update the title! 🙂
